lib/game_client/ogs/http_client.dart

```diff
@@ -0,0 +1,207 @@
+    int? apiVersion,
+  }) async {
+    final uri = _buildUri(path, queryParameters, apiVersion);
+
+    final response = await _httpClient.post(
+      uri,
+      headers: _buildHeaders(includeContentType: true, includeCsrf: true),
+      body: jsonEncode(data),
+    );
+
+    _checkResponse(response);
+    return response;
+  }
+
+  /// Makes a PUT request with JSON payload and returns the parsed JSON response
+  Future<Map<String, dynamic>> putJson(
+    String path,
+    Map<String, dynamic> data, {
+    Map<String, String>? queryParameters,
+    int? apiVersion,
+  }) async {
+    final response = await _put(path, data,
+        queryParameters: queryParameters, apiVersion: apiVersion);
+    return jsonDecode(response.body) as Map<String, dynamic>;
+  }
+
+  /// Makes a PUT request with JSON payload and returns the raw HTTP response
+  Future<http.Response> _put(
+    String path,
+    Map<String, dynamic> data, {
+    Map<String, String>? queryParameters,
+    int? apiVersion,
+  }) async {
+    final uri = _buildUri(path, queryParameters, apiVersion);
+
+    final response = await _httpClient.put(
+      uri,
+      headers: _buildHeaders(includeContentType: true, includeCsrf: true),
+      body: jsonEncode(data),
+    );
+
+    _checkResponse(response);
+    return response;
+  }
+
+  /// Builds the complete URI for a given path and query parameters
+  Uri _buildUri(
+      String path, Map<String, String>? queryParameters, int? apiVersion) {
+    // Use provided apiVersion, or default, or none
+    final version = apiVersion ?? defaultApiVersion;
+
+    String fullPath = version != null
+        ? '$serverUrl/api/v$version$path'
+        : '$serverUrl/api$path';
+
+    final uri = Uri.parse(fullPath);
+
+    if (queryParameters != null && queryParameters.isNotEmpty) {
+      return uri.replace(queryParameters: queryParameters);
+    }
+
+    return uri;
+  }
+
+  /// Builds the standard headers for requests
+  Map<String, String> _buildHeaders({
+    bool includeContentType = false,
+    bool includeCsrf = false,
+  }) {
+    final headers = <String, String>{
+      'User-Agent': _userAgent,
+    };
+
+    if (includeContentType) {
+      headers['Content-Type'] = 'application/json';
+    }
+
+    if (includeCsrf && _csrfToken != null) {
+      headers['X-CSRFToken'] = _csrfToken!;
+    }
+
+    return headers;
+  }
+
+  /// Checks the HTTP response and throws an exception for error status codes
+  void _checkResponse(http.Response response) {
+    if (response.statusCode < 200 || response.statusCode >= 300) {
+      throw HttpException(
+        'HTTP ${response.statusCode}: ${response.reasonPhrase}',
+        statusCode: response.statusCode,
+        responseBody: response.body,
+      );
+    }
+  }
+
+  /// Disposes of the underlying HTTP client
+  void dispose() {
+    _httpClient.close();
+  }
+}
+
+/// Exception thrown when an HTTP request fails
+class HttpException implements Exception {
```

Comment:
there's https://api.flutter.dev/flutter/dart-io/HttpException-class.html, but I guess yours also has additional info, which might be useful.

Reply:
makes sense :+1: